logger/custom_logger.go

```diff
@@ -27,19 +28,161 @@ const (
 )
 
 type CustomHandler struct {
-	h slog.Handler
-	b *bytes.Buffer
-	m *sync.Mutex
+	opts           Options
+	preformatted   []byte   // data from WithGroup and WithAttrs
+	unopenedGroups []string // groups from WithGroup that haven't been opened
+	indentLevel    int
+	mu             *sync.Mutex
+	out            io.Writer
 }
 
-// CustomHandler is a custom handler for logging records.
-func (h *CustomHandler) Handle(c context.Context, r slog.Record) error {
-	timestamp := time.Now().Format(time.RFC3339Nano)
+type Options struct {
+	// Level reports the minimum level to log.
+	// Levels with lower levels are discarded.
+	// If nil, the Handler uses [slog.LevelInfo].
+	Level slog.Leveler
+}
+
+func NewCustomHandler(out io.Writer, opts *Options) *CustomHandler {
+	h := &CustomHandler{out: out, mu: &sync.Mutex{}}
+	if opts != nil {
+		h.opts = *opts
+	}
+	if h.opts.Level == nil {
+		h.opts.Level = slog.LevelInfo
+	}
+	return h
+}
+
+// Handle is a custom handler for logging records.
+func (ch *CustomHandler) Handle(ctx context.Context, r slog.Record) error {
+	indentLevel := 0
+	bufp := allocBuf()
```

Comment:
is this pointer needed? I don't see it used anywhere in the function. You assign the value that it's pointing to to buf and that's what you are using throughout the rest of the function.

Reply:
`buffp` stands for buf Pool. See what `allocBuf()` returns at line 219.